kms/capi/capi.go

```diff
@@ -0,0 +1,754 @@
+		}
+	case "RSA":
+		buf, err := nCryptExportKey(kh, BCRYPT_RSAPUBLIC_BLOB)
+		if err != nil {
+			return nil, fmt.Errorf("failed to export %v public key: %w", algGroup, err)
+		}
+		pub, err = unmarshalRSA(buf)
+		if err != nil {
+			return nil, fmt.Errorf("failed to unmarshal %v public key: %w", algGroup, err)
+		}
+	default:
+		return nil, fmt.Errorf("unhandled algorithm group %v retrieved from key", algGroup)
+	}
+
+	return pub, nil
+}
+
+// New returns a new CAPIKMS.
+func New(ctx context.Context, opts apiv1.Options) (*CAPIKMS, error) {
+	providerName := "Microsoft Software Key Storage Provider"
+	pin := ""
+
+	if opts.URI != "" {
+		u, err := uri.ParseWithScheme(Scheme, opts.URI)
+		if err != nil {
+			return nil, err
+		}
+
+		if v := u.Get(ProviderNameArg); v != "" {
+			providerName = v
+		}
+
+		pin = u.Pin()
+	}
+
+	// TODO: a provider is not necessary for certificate functions, should we move this to the key and signing functions?
+	ph, err := nCryptOpenStorageProvider(providerName)
+	if err != nil {
+		return nil, fmt.Errorf("could not open nCrypt provider: %w", err)
+	}
+
+	return &CAPIKMS{
+		providerName:   providerName,
+		providerHandle: ph,
+		pin:            pin,
+	}, nil
+}
+
+func init() {
+	apiv1.Register(apiv1.CAPIKMS, func(ctx context.Context, opts apiv1.Options) (apiv1.KeyManager, error) {
+		return New(ctx, opts)
+	})
+}
+
+func (k *CAPIKMS) Close() error {
+	if k.providerHandle != 0 {
+		return nCryptFreeObject(k.providerHandle)
+	}
+
+	return nil
+}
+
+// CreateSigner returns a nce crypto.Signer that will sign using the key passed in via the URI.
+func (k *CAPIKMS) CreateSigner(req *apiv1.CreateSignerRequest) (crypto.Signer, error) {
+	u, err := uri.ParseWithScheme(Scheme, req.SigningKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse URI: %w", err)
+	}
+
+	var containerName string
+	if containerName = u.Get(ContainerNameArg); containerName == "" {
+		// generate a uuid for the container name
+		containerName, err = randutil.UUIDv4()
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate uuid: %w", err)
+		}
+	}
+
+	pinOrPass := u.Pin()
+	if pinOrPass == "" {
+		pinOrPass = k.pin
+	}
+
+	kh, err := nCryptOpenKey(k.providerHandle, containerName, 0, 0)
+	if err != nil {
+		return nil, fmt.Errorf("unable to open key: %w", err)
+	}
+
+	if pinOrPass != "" && k.providerName == ProviderMSSC {
+		err = nCryptSetProperty(kh, NCRYPT_PIN_PROPERTY, pinOrPass, 0)
+
+		if err != nil {
+			return nil, fmt.Errorf("unable to set key NCRYPT_PIN_PROPERTY: %w", err)
+		}
+	} else if pinOrPass != "" && k.providerName == ProviderMSPCP {
+		passHash, err := hashPasswordUTF16(pinOrPass)
+		if err != nil {
+			return nil, fmt.Errorf("unable to hash password: %w", err)
+		}
+
+		err = nCryptSetProperty(kh, NCRYPT_PCP_USAGE_AUTH_PROPERTY, passHash, 0)
+
+		if err != nil {
+			return nil, fmt.Errorf("unable to set key NCRYPT_PCP_USAGE_AUTH_PROPERTY: %w", err)
+		}
+	}
```

Comment:
It looks like that an `else` is missing here.

Reply:
if not supplied a UI prompt will be shown to the user, and at the moment we can only set the pin/pass for MSSC and MSPCP providers

Reply:
This is ok in most cases.